src/thor/bidirectional_astar.cc

```diff
@@ -917,129 +918,188 @@ void BidirectionalAStar::SetDestination(GraphReader& graphreader, const valhalla
 
 // Form the path from the adjacency list.
 std::vector<std::vector<PathInfo>> BidirectionalAStar::FormPath(GraphReader& graphreader,
-                                                                const valhalla::Options& options,
+                                                                const Options& options,
                                                                 const valhalla::Location& origin,
                                                                 const valhalla::Location& dest) {
-  // Get the indexes where the connection occurs.
-  uint32_t idx1 = edgestatus_forward_.Get(best_connection_.edgeid).index();
-  uint32_t idx2 = edgestatus_reverse_.Get(best_connection_.opp_edgeid).index();
 
-  // Metrics (TODO - more accurate cost)
-  uint32_t pathcost = edgelabels_forward_[idx1].cost().cost + edgelabels_reverse_[idx2].cost().cost;
-  LOG_DEBUG("path_cost::" + std::to_string(pathcost));
-  LOG_DEBUG("FormPath path_iterations::" + std::to_string(edgelabels_forward_.size()) + "," +
-            std::to_string(edgelabels_reverse_.size()));
+  // we need to figure out the maximum number of paths we could form here and
+  // if its more than 1 we need to sort them so we do the best first
+  size_t desired_paths = 1;
+  bool allow_alternates = options.has_alternates() && options.alternates() > 0;
 
-  // Work backwards on the forward path
-  std::vector<std::vector<PathInfo>> paths;
-  paths.emplace_back();
-  std::vector<PathInfo>& path = paths.back();
-  for (auto edgelabel_index = idx1; edgelabel_index != kInvalidLabel;
-       edgelabel_index = edgelabels_forward_[edgelabel_index].predecessor()) {
-    const BDEdgeLabel& edgelabel = edgelabels_forward_[edgelabel_index];
-    path.emplace_back(edgelabel.mode(), edgelabel.cost(), edgelabel.edgeid(), 0,
-                      edgelabel.restriction_idx(), edgelabel.transition_cost());
-
-    // Check if this is a ferry
-    if (edgelabel.use() == Use::kFerry) {
-      has_ferry_ = true;
-    }
+  LOG_DEBUG("Found connections before stretch filter: " + std::to_string(best_connections_.size()));
+
+  if (allow_alternates) {
+    desired_paths += options.alternates();
+    // Cull alternate paths longer than maximum stretch
+    // TODO: we should skip adding the connection at all if it's greater than stretch
+    filter_alternates_by_stretch(best_connections_);
```

Comment:
this is where we check the first viability critereon, removing any paths that are so much less optimal that we dont consider them